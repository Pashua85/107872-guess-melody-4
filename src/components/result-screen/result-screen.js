import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom';
import ActionCreator from '../../store/action-creator/action-creator';

const ResultScreen = (props) => {
  const {questions, mistakes, step, onAgainClick} = props;
  const questionsAmountString = getQuestionsAmountString(questions.length);
  const mistakesString = getMistakesString(mistakes);

  if (step === 0) {
    return <Redirect to="/dev-artist" />;
  }

  return (
    <section className="result">
      <div className="result__logo">
        <img src="img/melody-logo.png" alt="Угадай мелодию" width="186" height="83" />
      </div>
      <h2 className="result__title">Вы настоящий меломан!</h2>
      <p className="result__total">Вы ответили правильно на {questionsAmountString} и совершили {mistakesString}</p>
      <button
        className="replay"
        type="button"
        onClick={onAgainClick}
      >
        Сыграть ещё раз
      </button>
    </section>
  );
};

ResultScreen.propTypes = {
  questions: PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.string,
        questionText: PropTypes.string,
        tracks: PropTypes.arrayOf(
            PropTypes.shape({
              src: PropTypes.string.isRequired,
              artist: PropTypes.string,
              genre: PropTypes.string
            })
        ),
        answers: PropTypes.arrayOf(
            PropTypes.shape({
              picture: PropTypes.string,
              artist: PropTypes.string
            })
        )
      })
  ),
  mistakes: PropTypes.number.isRequired,
  step: PropTypes.number.isRequired,
  onAgainClick: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  questions: state.questions,
  mistakes: state.mistakes,
  step: state.step
});

const mapDispatchToProps = (dispatch) => ({
  onAgainClick: () => {
    dispatch(ActionCreator.restartGame());
  }
});

function getQuestionsAmountString(number) {
  const numberSting = number.toString(10);
  const lastDigit = number % 10;
  let questionsString = ``;

  if (number >= 11 && number <= 14) {
    questionsString = `вопросов`;
  } else if (lastDigit === 1 && number !== 11) {
    questionsString = `вопрос`;
  } else if (lastDigit >= 2 && lastDigit <= 4 && (number <= 11 || number >= 14)) {
    questionsString = `вопроса`;
  } else {
    questionsString = `вопросов`;
  }
  return `${numberSting} ${questionsString}`;
}

function getMistakesString(number) {
  const numberSting = number.toString(10);
  const lastDigit = number % 10;
  let mistakesString = ``;

  if (number >= 11 && number <= 14) {
    mistakesString = `ошибок`;
  } else if (lastDigit === 1 && number !== 11) {
    mistakesString = `ошибку`;
  } else if (lastDigit >= 2 && lastDigit <= 4 && (number <= 11 || number >= 14)) {
    mistakesString = `ошибки`;
  } else {
    mistakesString = `ошибок`;
  }
  return `${numberSting} ${mistakesString}`;
}

export default connect(mapStateToProps, mapDispatchToProps)(ResultScreen);
export {ResultScreen};

