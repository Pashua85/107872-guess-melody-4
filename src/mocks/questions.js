const AVATAR_URL = `https://api.adorable.io/avatars/17`;

export default [
  {
    type: `artist`,
    questionText: `Кто исполняет эту песню?`,
    song: {
      artist: `Gregory Porter`,
      src: `audio/gregory_porter_ hey_laura.mp3`
    },
    answers: [{
      picture: `${AVATAR_URL}/${Math.random()}`,
      artist: `Kurt Elling`
    }, {
      picture: `${AVATAR_URL}/${Math.random()}`,
      artist: `Gregory Porter`
    }, {
      picture: `${AVATAR_URL}/${Math.random()}`,
      artist: `Tony Bennett`
    }]
  },
  {
    type: `genre`,
    questionText: `Выберите музыку барокко`,
    genre: `baroque`,
    answers: [{
      src: `audio/handel_concerto_grosso_in_g_minor_allegro.mp3`,
      genre: `baroque`
    }, {
      src: `audio/haydn_symphony_no_6_finale.mp3`,
      genre: `classical`
    }, {
      src: `audio/franz_liszt_hungarian_rhapsody_no_11.mp3`,
      genre: `romantic`
    }, {
      src: `audio/philip_glass_string_quartet_no_2_company_movement_2`,
      genre: `minimalism`
    }]
  }
];
