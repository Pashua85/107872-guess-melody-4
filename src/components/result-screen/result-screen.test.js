import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {ResultScreen} from './result-screen';
import questions from '../../mocks/test-questions';

Enzyme.configure({
  adapter: new Adapter()
});

describe(`ResultScreen`, () => {
  it(`should render correctly`, () => {
    const wrapper = shallow(
        <ResultScreen
          questions={questions}
          mistakes={1}
          step={2}
          onAgainClick={() => {}}
        />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

