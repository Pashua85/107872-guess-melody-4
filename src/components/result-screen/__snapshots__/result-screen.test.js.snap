// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultScreen should render correctly 1`] = `ShallowWrapper {}`;
